README.md
# KZSwipeTableViewCell [![Build Status](https://travis-ci.org/k3zi/KZSwipeTableViewCell.svg?branch=master)](https://travis-ci.org/k3zi/KZSwipeTableViewCell)
Convenient UITableViewCell subclass that implements a swippable content to trigger actions (Swift Port of KZSwipeTableViewCell)
--------------------

<p align="center"><img src="https://raw.github.com/k3zi/KZSwipeTableViewCell/master/github-assets/mcswipe-front.png"/></p>

An effort to show how one would implement a UITableViewCell like the one we can see in the very well executed [Mailbox](http://www.mailboxapp.com/) iOS app.

##Preview
###Exit Mode
Swiping the cell should make it disappear. Convenient in destructive modes.

<p align="center"><img src="https://raw.github.com/k3zi/KZSwipeTableViewCell/master/github-assets/mcswipe-exit.gif"/></p>

###Switch Mode
The cell will bounce back after selecting a state, this allows you to keep the cell. Convenient to switch an option quickly.

<p align="center"><img src="https://raw.github.com/k3zi/KZSwipeTableViewCell/master/github-assets/mcswipe-switch.gif"/></p>

##Usage

```objc
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifier = @"Cell";

    KZSwipeTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];

    if (!cell) {
        cell = [[KZSwipeTableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:CellIdentifier];

        // Remove inset of iOS 7 separators.
        if ([cell respondsToSelector:@selector(setSeparatorInset:)]) {
            cell.separatorInset = UIEdgeInsetsZero;
        }

        [cell setSelectionStyle:UITableViewCellSelectionStyleGray];

        // Setting the background color of the cell.
        cell.contentView.backgroundColor = [UIColor whiteColor];
    }

    // Configuring the views and colors.
    UIView *checkView = [self viewWithImageName:@"check"];
    UIColor *greenColor = [UIColor colorWithRed:85.0 / 255.0 green:213.0 / 255.0 blue:80.0 / 255.0 alpha:1.0];

    UIView *crossView = [self viewWithImageName:@"cross"];
    UIColor *redColor = [UIColor colorWithRed:232.0 / 255.0 green:61.0 / 255.0 blue:14.0 / 255.0 alpha:1.0];

    UIView *clockView = [self viewWithImageName:@"clock"];
    UIColor *yellowColor = [UIColor colorWithRed:254.0 / 255.0 green:217.0 / 255.0 blue:56.0 / 255.0 alpha:1.0];

    UIView *listView = [self viewWithImageName:@"list"];
    UIColor *brownColor = [UIColor colorWithRed:206.0 / 255.0 green:149.0 / 255.0 blue:98.0 / 255.0 alpha:1.0];

    // Setting the default inactive state color to the tableView background color.
    [cell setDefaultColor:self.tableView.backgroundView.backgroundColor];

    [cell.textLabel setText:@"Switch Mode Cell"];
    [cell.detailTextLabel setText:@"Swipe to switch"];

    // Adding gestures per state basis.
    [cell setSwipeGestureWithView:checkView color:greenColor mode:KZSwipeTableViewCellModeSwitch state:KZSwipeTableViewCellState1 completionBlock:^(KZSwipeTableViewCell *cell, KZSwipeTableViewCellState state, KZSwipeTableViewCellMode mode) {
        NSLog(@"Did swipe \"Checkmark\" cell");
    }];

    [cell setSwipeGestureWithView:crossView color:redColor mode:KZSwipeTableViewCellModeSwitch state:KZSwipeTableViewCellState2 completionBlock:^(KZSwipeTableViewCell *cell, KZSwipeTableViewCellState state, KZSwipeTableViewCellMode mode) {
        NSLog(@"Did swipe \"Cross\" cell");
    }];

    [cell setSwipeGestureWithView:clockView color:yellowColor mode:KZSwipeTableViewCellModeSwitch state:KZSwipeTableViewCellState3 completionBlock:^(KZSwipeTableViewCell *cell, KZSwipeTableViewCellState state, KZSwipeTableViewCellMode mode) {
        NSLog(@"Did swipe \"Clock\" cell");
    }];

    [cell setSwipeGestureWithView:listView color:brownColor mode:KZSwipeTableViewCellModeSwitch state:KZSwipeTableViewCellState4 completionBlock:^(KZSwipeTableViewCell *cell, KZSwipeTableViewCellState state, KZSwipeTableViewCellMode mode) {
        NSLog(@"Did swipe \"List\" cell");
    }];

    return cell;
}
```

###Delegate

KZSwipeTableViewCell has a set of delegate methods in order to track the user behaviors. Take a look at the header file to be aware of all the methods provided by `KZSwipeTableViewCellDelegate`.

```objc
@interface MCTableViewController () <KZSwipeTableViewCellDelegate>
```

```objc
#pragma mark - KZSwipeTableViewCellDelegate

// Called when the user starts swiping the cell.
- (void)swipeTableViewCellDidStartSwiping:(KZSwipeTableViewCell *)cell;

// Called when the user ends swiping the cell.
- (void)swipeTableViewCellDidEndSwiping:(KZSwipeTableViewCell *)cell;

// Called during a swipe.
- (void)swipeTableViewCell:(KZSwipeTableViewCell *)cell didSwipeWithPercentage:(CGFloat)percentage;
```

###Cell Deletion
In `KZSwipeTableViewCellModeExit` mode you may want to delete the cell with a nice fading animation, the following lines will give you an idea how to execute it:

```objc
[cell setSwipeGestureWithView:crossView color:redColor mode:KZSwipeTableViewCellModeExit state:KZSwipeTableViewCellState2 completionBlock:^(KZSwipeTableViewCell *cell, KZSwipeTableViewCellState state, KZSwipeTableViewCellMode mode) {
    NSLog(@"Did swipe \"Cross\" cell");

    // Code to delete your cell...

}];
```

You can also ask for a confirmation before deleting a cell:

```objc

__weak MCTableViewController *weakSelf = self;

[cell setSwipeGestureWithView:crossView color:redColor mode:KZSwipeTableViewCellModeExit state:KZSwipeTableViewCellState1 completionBlock:^(KZSwipeTableViewCell *cell, KZSwipeTableViewCellState state, KZSwipeTableViewCellMode mode) {
    NSLog(@"Did swipe \"Cross\" cell");

    __strong MCTableViewController *strongSelf = weakSelf;
    strongSelf.cellToDelete = cell;

    UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:@"Delete?"
                                                        message:@"Are you sure your want to delete the cell?"
                                                       delegate:self
                                              cancelButtonTitle:@"No"
                                              otherButtonTitles:@"Yes", nil];
    [alertView show];
}];
```
Then handle the `UIAlertView` action:

```objc
#pragma mark - UIAlertViewDelegate

- (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex {

    // No
    if (buttonIndex == 0) {
        [_cellToDelete swipeToOriginWithCompletion:^{
            NSLog(@"Swiped back");
        }];
        _cellToDelete = nil;
    }

    // Yes
    else {
        // Code to delete your cell...
    }
}
```

There is also an example in the demo project, I recommend to take a look at it.

###Changing the trigger percentage
If the default trigger limits do not fit to your needs you can change them with the `firstTrigger` *(default: 25%)* and `secondTrigger` *(Default: 75%)* properties.

```objc
cell.settings.firstTrigger = 0.1;
cell.settings.secondTrigger = 0.5;
```

###Reseting cell position
It is possible to put the cell back to it's position when using the `KZSwipeTableViewCellModeExit` mode with the `-swipeToOriginWithCompletion:` method:

```objc
[cell swipeToOriginWithCompletion:^{
    NSLog(@"Cell swiped back!");
}];
```

## Installation

### CocoaPods

[CocoaPods](http://cocoapods.org) is a dependency manager for Cocoa projects.

```bash
$ gem install cocoapods
```

To integrate KZSwipeTableViewCell into your Xcode project using CocoaPods, specify it in your `Podfile`:

```ruby
pod 'KZSwipeTableViewCell', '~> 2.0'
```

Then, run the following command:

```bash
$ pod install
```

## Contact

Kesi Maduka

- http://kez.io
- me@kez.io

## Original Author

Ali Karagoz

- http://github.com/alikaragoz
- http://twitter.com/alikaragoz

## License

KZSwipeTableViewCell is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# CHANGELOG

## 2.0.0

* Adds property `settings: KZSwipeTableViewCellSettings` to customize the cell's actions
* New demo project (ported from MCSwipeTableViewCell)
* Ported over README file & assets 

## 1.0.0

Initial version.

KZSwipeTableViewCellExample/KZSwipeTableViewCellExample/AppDelegate.swift
//
//  AppDelegate.swift
//  KZSwipeTableViewCellExample
//
//  Created by Kesi Maduka on 2/9/16.
//  Copyright © 2016 Storm Edge Apps LLC. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(application: UIApplication, didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {
        // Override point for customization after application launch.
        
        window = UIWindow(frame: UIScreen.mainScreen().bounds)
        window?.rootViewController = UINavigationController(rootViewController: ViewController())
        window?.makeKeyAndVisible()
        
        return true
    }

    func applicationWillResignActive(application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



KZSwipeTableViewCellExample/KZSwipeTableViewCellExample/ViewController.swift
//
//  ViewController.swift
//  KZSwipeTableViewCellExample
//
//  Created by Kesi Maduka on 2/9/16.
//  Copyright © 2016 Storm Edge Apps LLC. All rights reserved.
//

import UIKit
import KZSwipeTableViewCell

class ViewController: UITableViewController {
    
    var numOfItems = 7
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        title = "Swipe Table View"
        if let nav = navigationController {
            nav.navigationBar.tintColor = UIColor.darkGrayColor()
        }
        navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .Refresh, target: self, action: Selector("reload"))
        
        let backgroundView = UIView(frame: view.bounds)
        backgroundView.backgroundColor = UIColor(white: 227/0/255.0, alpha: 1.0)
        self.tableView.backgroundView = backgroundView
        
        self.tableView.delegate = self
        self.tableView.dataSource = self
    }
    
    //MARK: TableView Data Source
    
    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return numOfItems
    }
    
    //MARK: TableView Delegate
    
    func configureCell(cell: KZSwipeTableViewCell, indexPath: NSIndexPath) {
        let checkView = KZSwipeTableViewCell.viewWithImageName("check")
        let greenColor = UIColor(red: 85.0/255.0, green: 213.0/255.0, blue: 80.0/255.0, alpha: 1.0)
        
        let crossView = KZSwipeTableViewCell.viewWithImageName("cross")
        let redColor = UIColor(red: 232.0/255.0, green: 61.0/255.0, blue: 14.0/255.0, alpha: 1.0)
        
        let clockView = KZSwipeTableViewCell.viewWithImageName("clock")
        let yellowColor = UIColor(red: 254.0/255.0, green: 217.0/255.0, blue: 56.0/255.0, alpha: 1.0)
        
        let listView = KZSwipeTableViewCell.viewWithImageName("list")
        let brownColor = UIColor(red: 206.0/255.0, green: 149.0/255.0, blue: 98.0/255.0, alpha: 1.0)
        
        if let bgView = self.tableView.backgroundView {
            if let bgColor = bgView.backgroundColor {
                cell.settings.defaultColor = bgColor
            }
        }
        
        if indexPath.row % numOfItems == 0 {
            cell.textLabel?.text = "Switch Mode Cell"
            cell.detailTextLabel?.text = "Swipe to switch"
            
            cell.setSwipeGestureWith(checkView, color: greenColor, mode: .Switch, state: .State1, completionBlock: { (cell, state, mode) -> Void in
                print("Did swipe Checkmark cell")
            })
            
            cell.setSwipeGestureWith(crossView, color: redColor, mode: .Switch, state: .State2, completionBlock: { (cell, state, mode) -> Void in
                print("Did swipe Cross cell")
            })
            
            cell.setSwipeGestureWith(clockView, color: yellowColor, mode: .Switch, state: .State3, completionBlock: { (cell, state, mode) -> Void in
                print("Did swipe Clock cell")
            })
            
            cell.setSwipeGestureWith(listView, color: brownColor, mode: .Switch, state: .State4, completionBlock: { (cell, state, mode) -> Void in
                print("Did swipe List cell")
            })
        } else if indexPath.row % numOfItems == 1 {
            cell.textLabel?.text = "Exit Mode Cell"
            cell.detailTextLabel?.text = "Swipe to delete"
            
            cell.setSwipeGestureWith(crossView, color: redColor, mode: .Switch, state: .State1, completionBlock: { (cell, state, mode) -> Void in
                print("Did swipe Cross cell")
                
                self.deleteCell(cell)
            })
        } else if indexPath.row % numOfItems == 2 {
            cell.textLabel?.text = "Mixed Mode Cell"
            cell.detailTextLabel?.text = "Swipe to switch or delete"
            cell.settings.shouldAnimateIcons = true
            
            cell.setSwipeGestureWith(checkView, color: greenColor, mode: .Switch, state: .State1, completionBlock: { (cell, state, mode) -> Void in
                print("Did swipe Checkmark cell")
            })
            
            cell.setSwipeGestureWith(crossView, color: redColor, mode: .Exit, state: .State2, completionBlock: { (cell, state, mode) -> Void in
                print("Did swipe Cross cell")
                
                self.deleteCell(cell)
            })
        } else if indexPath.row % numOfItems == 3 {
            cell.textLabel?.text = "Un-animated Icons"
            cell.detailTextLabel?.text = "Swipe"
            cell.settings.shouldAnimateIcons = false
            
            cell.setSwipeGestureWith(checkView, color: greenColor, mode: .Switch, state: .State1, completionBlock: { (cell, state, mode) -> Void in
                print("Did swipe Checkmark cell")
            })
            
            cell.setSwipeGestureWith(crossView, color: redColor, mode: .Exit, state: .State2, completionBlock: { (cell, state, mode) -> Void in
                print("Did swipe Cross cell")
                
                self.deleteCell(cell)
            })
        } else if indexPath.row % numOfItems == 4 {
            cell.textLabel?.text = "Right swipe only"
            cell.detailTextLabel?.text = "Swipe"
            
            cell.setSwipeGestureWith(clockView, color: yellowColor, mode: .Switch, state: .State3, completionBlock: { (cell, state, mode) -> Void in
                print("Did swipe Clock cell")
            })
            
            cell.setSwipeGestureWith(listView, color: brownColor, mode: .Switch, state: .State4, completionBlock: { (cell, state, mode) -> Void in
                print("Did swipe List cell")
            })
        } else if indexPath.row % numOfItems == 5 {
            cell.textLabel?.text = "Small triggers"
            cell.detailTextLabel?.text = "Using 10% and 50%"
            cell.settings.firstTrigger = 0.1
            cell.settings.secondTrigger = 0.5
            
            cell.setSwipeGestureWith(checkView, color: greenColor, mode: .Switch, state: .State1, completionBlock: { (cell, state, mode) -> Void in
                print("Did swipe Checkmark cell")
            })
            
            cell.setSwipeGestureWith(crossView, color: redColor, mode: .Exit, state: .State2, completionBlock: { (cell, state, mode) -> Void in
                print("Did swipe Cross cell")
                
                self.deleteCell(cell)
            })
        } else if indexPath.row % numOfItems == 6 {
            cell.textLabel?.text = "Exit Mode Cell + Confirmation"
            cell.detailTextLabel?.text = "Swipe to delete"
            
            cell.setSwipeGestureWith(crossView, color: redColor, mode: .Exit, state: .State1, completionBlock: { (cell, state, mode) -> Void in
                print("Did swipe Cross cell")
                
                let alert = UIAlertController(title: "Delete?", message: "Are you sure your want to delete the cell?", preferredStyle: .Alert)
                alert.addAction(UIAlertAction(title: "No", style: .Cancel, handler: { (action) -> Void in
                    cell.swipeToOriginWithCompletion({ () -> Void in
                        print("swiped back")
                    })
                }))
                alert.addAction(UIAlertAction(title: "Yes", style: .Destructive, handler: { (action) -> Void in
                    self.deleteCell(cell)
                }))
                self.presentViewController(alert, animated: true, completion: nil)
            })
        }
    }
    
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        
        var cell = tableView.dequeueReusableCellWithIdentifier("Cell")
        
        if cell == nil {
            cell = KZSwipeTableViewCell(style: .Subtitle, reuseIdentifier: "Cell")
        }
        
        if let cell = cell as? KZSwipeTableViewCell {
            configureCell(cell, indexPath: indexPath)
            return cell
        }
        
        return UITableViewCell()
    }
    
    override func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        return 70.0
    }
    
    func reload() {
        numOfItems = 7
        tableView.reloadSections(NSIndexSet(index: 0), withRowAnimation: .Fade)
    }
    
    func deleteCell(cell: KZSwipeTableViewCell) {
        numOfItems--;
        if let indexPath = tableView.indexPathForCell(cell) {
            tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation: .Fade)
        }
    }
}
